src/geom/util/GeometryFixer.cpp

```diff
@@ -0,0 +1,337 @@
+    if (typeId == GEOS_LINESTRING)         return fixLineString(static_cast<const LineString*>(geom));
+    if (typeId == GEOS_POLYGON)            return fixPolygon(static_cast<const Polygon*>(geom));
+    if (typeId == GEOS_MULTIPOINT)         return fixMultiPoint(static_cast<const MultiPoint*>(geom));
+    if (typeId == GEOS_MULTILINESTRING)    return fixMultiLineString(static_cast<const MultiLineString*>(geom));
+    if (typeId == GEOS_MULTIPOLYGON)       return fixMultiPolygon(static_cast<const MultiPolygon*>(geom));
+    if (typeId == GEOS_GEOMETRYCOLLECTION) return fixCollection(static_cast<const GeometryCollection*>(geom));
+    throw geos::util::UnsupportedOperationException("GeometryFixer::getResult called on unknown geometry type");
+}
+
+/* private */
+std::unique_ptr<Point>
+GeometryFixer::fixPoint(const Point* p_geom)
+{
+    std::unique_ptr<Point> pt = fixPointElement(p_geom);
+    if (pt == nullptr)
+        return factory->createPoint();
+    else
+        return pt;
+}
+
+/* private */
+std::unique_ptr<Point>
+GeometryFixer::fixPointElement(const Point* p_geom)
+{
+    if (p_geom->isEmpty() || ! isValidPoint(p_geom)) {
+        return std::unique_ptr<Point>(nullptr);
+    }
+    return p_geom->clone();
+}
+
+/* private static */
+bool
+GeometryFixer::isValidPoint(const Point* pt)
+{
+    const Coordinate* p = pt->getCoordinate();
+    return p->isValid();
+}
+
+/* private */
+std::unique_ptr<Geometry>
+GeometryFixer::fixMultiPoint(const MultiPoint* p_geom)
+{
+    std::vector<std::unique_ptr<Point>> pts;
+    for (std::size_t i = 0; i < p_geom->getNumGeometries(); i++) {
+        const Point* pt = p_geom->getGeometryN(i);
+        if (pt->isEmpty()) continue;
+        std::unique_ptr<Point> fixPt = fixPointElement(pt);
+        if (fixPt != nullptr) {
+            pts.emplace_back(fixPt.release());
+        }
+    }
+    return factory->createMultiPoint(std::move(pts));
+}
+
+/* private */
+std::unique_ptr<Geometry>
+GeometryFixer::fixLinearRing(const LinearRing* p_geom)
+{
+    std::unique_ptr<Geometry> fix = fixLinearRingElement(p_geom);
+    if (fix == nullptr)
+        return factory->createLinearRing();
+    else
+        return fix;
+}
+
+/* private */
+std::unique_ptr<Geometry>
+GeometryFixer::fixLinearRingElement(const LinearRing* p_geom)
+{
+    if (p_geom->isEmpty())
+        return std::unique_ptr<Geometry>(nullptr);
+
+    auto ptsFix = RepeatedPointRemover::removeRepeatedAndInvalidPoints(p_geom->getCoordinatesRO());
+    auto ptsFixSz = ptsFix->size();
+    if (isKeepCollapsed) {
+        if (ptsFixSz == 1) {
+            const Coordinate& c = ptsFix->getAt(0);
+            std::unique_ptr<Geometry> pt(factory->createPoint(c));
+            return pt;
+        }
+        if (ptsFixSz > 1 && ptsFixSz <= 3) {
+            return factory->createLineString(std::move(ptsFix));
+        }
+    }
+    //--- too short to be a valid ring
+    if (ptsFixSz <= 3)
+        return std::unique_ptr<Geometry>(nullptr);
```

Comment:
```suggestion
        return nullptr;
```
similar instances below, but I won't spam this review with them :-)